### PR TITLE
feat: record Success-Fail telemetry event on job attachments upload and download.

### PR DESCRIPTION
**Pull request by leongdl**
### What was the problem/requirement? (What/Why)
- deadline cloud had a regression when 0.48.2 release was pushed to Pypi. 
- The regression was root caused to Job Attachments download code paths after enabling boto3 with CRT.
- The client side lacked telemetry of different upload / download use cases that can be good signals of a regression.

### What was the solution? (How)
- Add a try-caught wrapper decorator that can be used to emit simple success or fail telemetries.
- Add the decorator to both Job Submission (CLI / GUI) code paths and to Job output download code paths.

### What is the impact of this change?
- Capture more data about how the component is behaving after release.

### How was this change tested?
- hatch run build
- hatch run fmt
- hatch run test
```
------------------------------------------------------------------------------------------------------------------------
TOTAL                                                                         4808    672   2110    204    84%
Coverage HTML written to dir build/coverage
Coverage XML written to file build/coverage/coverage.xml

Required test coverage of 80.0% reached. Total coverage: 84.42%
```
- hatch run integ:test
```
scheduling tests via LoadScheduling

test/integ/deadline_job_attachments/test_job_attachments.py::test_upload_input_files_all_assets_in_cas[2023-03-03] 
[gw0] [  7%] PASSED test/integ/deadline_job_attachments/test_job_attachments.py::test_upload_input_files_all_assets_in_cas[2023-03-03] 
test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_outputs_no_job_attachment_settings_in_job[2023-03-03] 
[gw0] [ 14%] PASSED test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_outputs_no_job_attachment_settings_in_job[2023-03-03] 
test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_outputs_no_job_attachment_s3_settings[2023-03-03] 
[gw0] [ 21%] PASSED test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_outputs_no_job_attachment_s3_settings[2023-03-03] 
test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_outputs_with_symlink[2023-03-03] 
[gw0] [ 28%] PASSED test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_outputs_with_symlink[2023-03-03] 
test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_inputs_with_step_dependencies[2023-03-03] 
[gw0] [ 35%] PASSED test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_inputs_with_step_dependencies[2023-03-03] 
test/integ/deadline_job_attachments/test_job_attachments.py::test_download_outputs_with_job_id_step_id_task_id_and_download_directory[2023-03-03] 
[gw0] [ 42%] PASSED test/integ/deadline_job_attachments/test_job_attachments.py::test_download_outputs_with_job_id_step_id_task_id_and_download_directory[2023-03-03] 
test/integ/deadline_job_attachments/test_job_attachments.py::test_download_outputs_with_job_id_step_id_and_download_directory[2023-03-03] 
[gw0] [ 50%] PASSED test/integ/deadline_job_attachments/test_job_attachments.py::test_download_outputs_with_job_id_step_id_and_download_directory[2023-03-03] 
test/integ/deadline_job_attachments/test_job_attachments.py::test_download_outputs_with_job_id_and_download_directory[2023-03-03] 
[gw0] [ 57%] PASSED test/integ/deadline_job_attachments/test_job_attachments.py::test_download_outputs_with_job_id_and_download_directory[2023-03-03] 
test/integ/deadline_job_attachments/test_job_attachments.py::test_upload_input_files_no_download_paths[2023-03-03] 
[gw0] [ 64%] PASSED test/integ/deadline_job_attachments/test_job_attachments.py::test_upload_input_files_no_download_paths[2023-03-03] 
test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_inputs_no_inputs[2023-03-03] 
[gw0] [ 71%] PASSED test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_inputs_no_inputs[2023-03-03] 
test/integ/deadline_job_attachments/test_job_attachments.py::test_upload_bucket_wrong_account[2023-03-03] 
[gw0] [ 78%] PASSED test/integ/deadline_job_attachments/test_job_attachments.py::test_upload_bucket_wrong_account[2023-03-03] 
test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_inputs_bucket_wrong_account[2023-03-03] 
[gw0] [ 85%] PASSED test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_inputs_bucket_wrong_account[2023-03-03] 
test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_outputs_bucket_wrong_account[2023-03-03] 
[gw0] [ 92%] PASSED test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_outputs_bucket_wrong_account[2023-03-03] 
test/integ/deadline_job_attachments/test_job_attachments.py::test_download_outputs_bucket_wrong_account[2023-03-03] 
[gw0] [100%] PASSED test/integ/deadline_job_attachments/test_job_attachments.py::test_download_outputs_bucket_wrong_account[2023-03-03] 

================================================================ slowest 5 durations =================================================================
3.04s setup    test/integ/deadline_job_attachments/test_job_attachments.py::test_upload_input_files_all_assets_in_cas[2023-03-03]
1.56s setup    test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_inputs_with_step_dependencies[2023-03-03]
1.36s teardown test/integ/deadline_job_attachments/test_job_attachments.py::test_download_outputs_bucket_wrong_account[2023-03-03]
1.25s setup    test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_outputs_no_job_attachment_settings_in_job[2023-03-03]
1.18s call     test/integ/deadline_job_attachments/test_job_attachments.py::test_sync_outputs_no_job_attachment_settings_in_job[2023-03-03]
================================================================ 14 passed in 17.45s =================================================================

```
Manual testing:
- Performed a Job submission using a blender bundle I have. Assets were hashed and uploaded.
- Downloaded the output of the blender job I submitted. The job outputs were downloaded correctly.

### Was this change documented?
- Not applicable, this is not changing any contracts.

### Is this a breaking change?
- No, this is internal.
----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/deadline/client/api/_telemetry.py [hunk 1]
```diff
@@ -349,3 +353,35 @@ def get_deadline_cloud_library_telemetry_client(
     :return: Telemetry client to make requests with.
     """
     return get_telemetry_client("deadline-cloud-library", version, config=config)
+
+
+def record_success_fail_telemetry_event(**decorator_kwargs: Dict[str, Any]) -> Callable[..., F]:
```
Comment (leongdl) [inline src/deadline/client/api/_telemetry.py:358]: I made this a decorator so we can tag any code path to emit success / fail from now on.

**Review by crowecawcaw — DISMISSED** [review]

File: src/deadline/client/api/_telemetry.py [hunk 2]
```diff
@@ -349,3 +353,35 @@ def get_deadline_cloud_library_telemetry_client(
...
+    :param ** Python variable arguments. See https://docs.python.org/3/glossary.html#term-parameter.
+    """
+
+    def inner(function: F) -> F:
+        def wrapper(*args: List[Any], **kwargs: Dict[str, Any]) -> Any:
+            """
+            Wrapper to try-catch a function for telemetry
+            :param * Python variable argument. See https://docs.python.org/3/glossary.html#term-parameter
+            :param ** Python variable argument. See https://docs.python.org/3/glossary.html#term-parameter
+            :
```
Comment (benl-2023) [inline src/deadline/client/api/_telemetry.py]: nit: looks like an extra colon?

**Review by YutongLi291 — DISMISSED** [review]

File: src/deadline/client/cli/_groups/job_group.py [hunk 3]
```diff
@@ -396,6 +396,19 @@ def _download_job_output(
     # in Job Attachments library can be increased (currently using default number, 10, which
     # makes it keep logging urllib3 warning messages when downloading large files)
     with _modified_logging_level(logging.getLogger("urllib3"), logging.ERROR):
+
+        @api.record_success_fail_telemetry_event(metric_name="download_job_output")  # type: ignore
```
Comment (benl-2023) [inline src/deadline/client/cli/_groups/job_group.py:400]: Looks like here, and other uses of this decorator we have to disable type checking. Is there alternative to this?

File: src/deadline/client/api/_telemetry.py (hunk 2 above)
Comment (leongdl) [reply src/deadline/client/api/_telemetry.py]: Good point- let me delete it.

File: src/deadline/client/cli/_groups/job_group.py (hunk 3 above)
Comment (leongdl) [reply src/deadline/client/cli/_groups/job_group.py:400]: No - it is due to mac mypy issues. It passes on all other operating systems.

**Review by benl-2023 — APPROVED** [review]

**Review by crowecawcaw — APPROVED** [review]

**Pull request merged**